### DISCUSSION TITLE
What's the best way to do code-splitting with Chakra?

### DISCUSSION THREAD

**Discussion by vonwao** [chakra-ui/chakra-ui#2291] [Q&A] [answered]
I first assumed that if I import using the {} syntax from @chakra-ui/core that webpack would be smart enough to split individual chakra components into separate bundles.  

(e.g. import { ChakraProvider, extendTheme } from "@chakra-ui/core")

But that doesn't seem to be the case.  It seems that as soon as I import anything from chakra, everything ends up in that bundle (because my main bundle is like 500K and my other bundle is like 3K). 

Thanks, any tips appreciated!

**Comment by ljosberinn** [comment, marked as answer]
What bundling tool do you use and which setup? Chakra treeshakes just fine, here's a screenshot of the landing page of a project of mine, as you can see only those that I actually use are imported. This is the default without any webpack tinkering in latest nextjs + webpack 5.
![image](https://user-images.githubusercontent.com/29307652/96945908-43a3c180-14df-11eb-967b-8782a5b0e20f.png)

**Reply by vonwao** [reply]
@ljosberinn  I set up with create-react-app (using webpack 4.42.0).   This should be doing tree-shaking?  Thanks.

**Reply by vonwao** [reply]
Below is what my bundle looks like.  So I don't think it's doing either tree shaking or code-splitting correctly, because there's definitely a bunch of stuff in there that I'm a) not using at all or b) should be in another bundle

![image](https://user-images.githubusercontent.com/377000/97025858-9eacd700-1526-11eb-8cf7-d146412d0db7.png)

**Reply by ljosberinn** [reply]
Yeah create-react-app should treeshake out of the box. FWIW, some cross-dependencies are to be expected, because let's say `Menu` is used, `popper` will be pulled in, so it's generally expected to see a little more than you may expect.
Is this a repo I can look at? From here I can't tell which dependency in there _should_ or _should not_ be there because there's rarely a reason to doubt the bundler.

**Reply by vonwao** [reply]
Thanks @ljosberinn .   I can't share this repo.  I will try to provide a real example when I have a chance to.   My main question was that if I want to code split an app, if that will work with chakra with the imports I have above, but I guess I'll need to provide a real example.

**Reply by ljosberinn** [reply]
Generally speaking there shouldn't be any issues. At least we aren't aware of them, and I'm having an eye on bundle size personally and haven't noticed anything unusual at least.

**Reply by SheaBelsky** [reply]
@vonwao What tool are you using to get a bundle analysis diagram like that?

**Reply by verystack** [reply]
@SheaBelsky that's tree view from PageSpeed Insights. 

Wondering what difference is tree-shake wise between importing from @chakra-ui/react, and individual chakra packages?

**Reply by ljosberinn** [reply]
> Wondering what difference is tree-shake wise between importing from @chakra-ui/react, and individual chakra packages?

none
